src/memtest.rs

```diff
@@ -0,0 +1,378 @@
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+pub unsafe fn test_block_seq(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_ms: usize,
+) -> Result<MemtestOutcome, MemtestError> {
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let mut timeout_checker = TimeoutChecker::new(Instant::now(), timeout_ms, half_count * 2 * 256);
+
+    for i in 0..=255 {
+        let mut val: usize = 0;
+        write_bytes(&mut val, i, size_of::<usize>() / size_of::<u8>());
+        for j in 0..half_count {
+            timeout_checker.check()?;
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, &mut timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+impl TimeoutChecker {
+    fn new(start_time: Instant, timeout_ms: usize, total_iter: usize) -> TimeoutChecker {
+        TimeoutChecker {
+            start_time,
+            timeout_ms,
+            total_iter,
+            completed_iter: 0,
+            checkpoint: 1,
+            num_checks_completed: 0,
+            // TODO: Choice of starting interval is arbitrary for now.
+            checking_interval_ns: 1000.0,
+        }
+    }
+
+    /// This function should be called in every iteration of a memtest
+    ///
+    /// To minimize overhead, the time is only checked at specific checkpoints.
+    /// The algorithm makes a prediction of how many iterations will be done during `checking_interval_ns`
+    /// and sets the checkpoint based on the prediction.
+    ///
+    /// The algorithm also determines whether it is likely that the test will be completed
+    /// based on `work_progress` and `time_progress`.
+    /// If it is likely that the test will be completed, `checking_interval_ns` is scaled up to be more
+    /// lenient and reduce overhead.
+    fn check(&mut self) -> Result<(), MemtestError> {
+        if self.completed_iter < self.checkpoint {
+            self.completed_iter += 1;
+            return Ok(());
+        }
+
+        let elapsed = self.start_time.elapsed();
+        if elapsed > Duration::from_millis(self.timeout_ms as u64) {
+            return Err(MemtestError::Timeout);
+        }
+
+        let work_progress = self.completed_iter as f64 / self.total_iter as f64;
+        let time_progress = elapsed.as_millis() as f64 / self.timeout_ms as f64;
+        // TODO: Consider having a max for `checking_interval_ns` to have a reasonable timeout guarantee
+        if work_progress > time_progress {
+            self.checking_interval_ns *= 2.0;
+        }
+
+        let avg_iter_time_ns = elapsed.as_nanos() as f64 / self.completed_iter as f64;
+        let iter_per_interval = self.checking_interval_ns / avg_iter_time_ns;
+        self.checkpoint = self.completed_iter + iter_per_interval as usize;
+
+        self.num_checks_completed += 1;
+        self.completed_iter += 1;
+        Ok(())
+    }
+}
```

Comment:
Implemented this timeout checking function to minimize overhead (each iteration in a test takes ~1ns, while checking the time takes ~30ns). It also tries to further reduce overhead by reducing the number of checks to do if it is optimistic about the tests completion before timeout.
The irony is that if it seems that it is unlikely for the test to complete before timeout, then it has to keep the same interval of checking, which means more overhead and even less chance to complete before timeout. Not sure if there's a better way to handle this.